packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -2171,18 +2172,24 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
     }
 
     caretRect = caretRect.shift(_getPixelPerfectCursorOffset(caretRect));
+    _currentCaretRect = caretRect;
+
+    if (!_showCursor.value)
+      return;
 
     if (cursorRadius == null) {
       canvas.drawRect(caretRect, paint);
     } else {
       final RRect caretRRect = RRect.fromRectAndRadius(caretRect, cursorRadius!);
       canvas.drawRRect(caretRRect, paint);
     }
+  }
 
-    if (caretRect != _lastCaretRect) {
-      _lastCaretRect = caretRect;
+  void _updateCaretRect() {
+    if (_currentCaretRect != _lastCaretRect) {
+      _lastCaretRect = _currentCaretRect;
       if (onCaretChanged != null)
-        onCaretChanged!(caretRect);
+        onCaretChanged!(_currentCaretRect);
```

Comment:
The root cause of the issues is that the `caretRect` did not update to `EditableTextState._showCaretOnScreen`.

Reply:
What's the reason for splitting this stuff into a new method?  Now you must call both `_paintCaretIfNeeded` and `_updateCaretRect` back-to-back, which I worry someone might forget to do in the future.